### PR TITLE
feat(compiler): close the five zero-global-cost native gaps (decorators, os.path, mixed isinstance, closures, generators)

### LINKED ISSUE #7588: Native pathway: close the five zero-global-cost language gaps (os.path stragglers, mixed-tuple isinstance, static decorators, generators, closure captures)
# Native pathway: close the five zero-global-cost language gaps (os.path stragglers, mixed-tuple isinstance, static decorators, generators, closure captures)

## Motivation and governing principle

Auditing what remains unsupported on the native pathway, five gaps stand out because they can all be lowered with **zero cost to programs that do not use them**: no object-representation changes, no new global runtime state, no deoptimization of field access or dispatch. Every one compiles away at the use site. This distinguishes them from the gaps we should keep excluding on principle (dynamic `getattr` / monkey-patching, metaclasses / dynamic `class`), whose support would tax every field load in every program.

The governing constraint for everything below:

> A feature may land on the native pathway only if its entire cost is confined to the code that uses it. Fixed-offset field loads, static vtables, and non-boxed locals stay untouched.

Each workstream below states current behavior (with code references), a lowering sketch intended to be the proper long-term design rather than a stopgap, semantics notes, and acceptance tests. The workstreams are independently landable.

---

## Workstream A: `os.path.split` / `splitext` / `abspath` / `normpath`

**Current state.** `NaIRGenPass._emit_os_call` (`jac/jaclang/compiler/passes/native/na_ir_gen_pass.impl/os.impl.jac`) lowers `basename`, `dirname`, `exists`, `join`, `realpath`, `getcwd`, `getenv`, `getpid`, `isdir`, `isfile`, `mkdir`, `makedirs`, `rename`, `system` via libc externs plus `_rc_strdup`. The four named functions are absent. Note that `realpath` is **not** a substitute for `abspath`: it resolves symlinks and touches the filesystem; `abspath` must be a pure string computation over `getcwd()`.

**Sketch.** These four are pure string algorithms, which suggests they should not grow as bespoke IR emission inside `_emit_os_call` at all. The proper long-term shape is the pattern already proven by the OSP kernel (`jac/jaclang/jac0core/osp_kernel.jac`): implement the path-string algorithms once in a small portable Jac kernel (e.g. `jac0core/path_kernel.jac`) as pure `str -> str` / `str -> tuple[str, str]` functions, compiled by every backend. That yields cross-backend equality by construction instead of by test discipline, and `_emit_os_call` keeps only true syscall shims.

- `split(p)` / `splitext(p)`: single reverse scan; return `tuple[str, str]` (native tuples already lower).
- `normpath(p)`: the standard segment-stack algorithm (collapse `//`, drop `.`, fold `..` except at root / leading positions).
- `abspath(p)`: `normpath(join(getcwd(), p))` when `p` is relative; `getcwd` already lowers natively.

**Semantics notes.**
- Posix semantics first. The PE linker exists (`pe_linker.jac`), so parametrize the separator through one kernel-level constant when a Windows target is active, but an `ntpath`-faithful variant (drive letters, UNC) is out of scope here.
- `splitext` corner cases must match CPython: leading-dot files (`".bashrc"` has no extension), multiple dots (`"a.tar.gz"` splits at the last), trailing dots.

**Acceptance.** Extend the existing `prim_ospath.jac` equivalence fixture (`jac/jaclang/compiler/tests/fixtures/`) under `run_with_both(..., require=["na"])` with: empty string, `"."`, `".."`, trailing slashes, duplicated separators, dotfiles, `normpath("a//b/../c")`, absolute and relative inputs to `abspath` (with the test controlling cwd).

---

## Workstream B: `isinstance` over mixed builtin + archetype tuples

**Current state.** `NaIRGenPass._emit_isinstance` (`jac/jaclang/compiler/passes/native/na_ir_gen_pass.impl/calls.impl.jac`) resolves each tuple arm to JacVal tags via `_jacval_tags_for_type` (`int` [0,7], `float` [1], `str` [2], `list` [3], `dict` [4], `set` [5], `bool` [7], `NoneType` [8]). All-builtin tuples lower to an OR-chain of tag compares; all-archetype targets go through `_emit_isinstance_user` (vtable / MRO scan, landed in #6932). A tuple mixing the two families deliberately raises structured E5092 rather than silently dropping the archetype arm (review follow-up from #6421; guarded by the test `isinstance mixed tuple unsupported type errors` and fixture `na_isinstance_mixed_tuple.na.jac`).

**Sketch.** Both mechanisms exist; the fix is emitting them as a guarded disjunction on one boxed operand:

1. Box once via `_box_to_jacval`; extract the tag.
2. Builtin arms: existing OR-chain of tag compares.
3. Archetype arms: branch taken only when `tag == 6` (the user-struct pointer tag assigned in `_box_to_jacval`). Inside that branch, `inttoptr` the payload, load the vtable pointer at slot 0, and compare pointer equality against the target vtable set. The target set computation already exists in `_emit_isinstance_user`: scan `self.vtable_globals` for candidates whose MRO (`_mro_of`) contains any named target, so subclass instances match base-class arms.
4. OR the two sides; keep the existing single RC release of an owned operand.

The short-circuiting branch in (3) is a correctness requirement, not an optimization: the vtable load must be unreachable when the payload is not an object pointer.

One generalization is needed: `_emit_isinstance_user` currently requires a statically known struct name and bails for vtable-less types. For an `any`-typed operand the static struct is unknown, and an archetype target with no hierarchy may have no vtable today. The long-term-correct resolution is to force vtable materialization for any archetype named as an `isinstance` target (consistent with the direction of `class_name.na.jac`, where hierarchy participation already flags types into vtables), rather than inventing a parallel type-id side channel.

**Acceptance.** Flip the E5092 test into positive equivalence tests: `isinstance(x, (int, MyNode))` over `any`-boxed values of every tag; a subclass instance matching a base arm inside a mixed tuple; negative cases (float against `(int, MyNode)`); `None` handling. Keep one negative test asserting no silent arm-dropping for genuinely unresolvable names.

---

## Workstream C: decorators (static subset), guardrail first

**Current state, including a latent correctness bug.** There is no decorator handling anywhere under `jac/jaclang/compiler/passes/native/`, and decorators appear in neither `NATIVE_REJECT_NODES` nor the walk in `native_capability_violations` (`jac/jaclang/compiler/passes/main/capability_check_pass.jac`, `.../impl/capability_check_pass.impl.jac`). A decorated function in a `.na.jac` module is therefore **silently compiled without its decorator**: the undecorated body runs. That is the only item in this issue that is a live correctness hole rather than a missing feature.

**Phase C0 (immediate, independent of the rest).** Reject decorated abilities in native modules with E5090 in `native_capability_violations` (the `Ability.decorators` / `ImplDef.decorators` fields are already on the nodes), with help text pointing at the Python-side escape hatch. Add a matching guardrail test alongside the existing ones in `jac/tests/compiler/passes/native/test_native_feature_guardrail.jac`.

**Phase C1 (static subset).** Support decorators whose value is statically resolvable to a native function of shape `(fn) -> fn`:

1. Lower the decorated function body as a private raw symbol `__jac_raw_f`.
2. At module glob-init (matching Python's import-time decoration order, bottom-up for stacked decorators), call the decorator with the raw function value and store the result in a private module slot.
3. Call sites of `f` load the slot and call indirectly: one indirect call, only for decorated functions.

The two required ingredients already lower today: function-typed values / fields (`callable_field.na.jac`) and wrapper closures with bound state (the lambda capture machinery in `_codegen_lambda`, `jac/.../na_ir_gen_pass.impl/expr.impl.jac`). Registration-style decorators that return their argument unchanged get their direct calls restored by a peephole (slot provably initialized to the raw symbol), so the common marker-decorator case costs nothing.

**Scope.** Module-level functions only; method decorators keep the E5090 guardrail until the vtable interplay is designed. Non-goals: dynamically computed decorator expressions, class decorators.

**Acceptance.** Equivalence tests for: a wrapping decorator (arg/result transformation), a stacked pair (order), a registration decorator (side effect at glob-init plus unchanged call), and a guardrail test that method decoration still rejects cleanly.

---

## Workstream D: generators / `yield` (state-machine lowering)

**Current state.** `YieldExpr` is a hard reject in `NATIVE_REJECT_NODES` (`capability_check_pass.jac:35`), plus the guardrail test. Meanwhile #6403 landed a complete first-class iterator ABI: `JacIter { next_fn, state }` with per-source state structs and an out-slot `next()` protocol raising StopIteration past the end (`jac/.../na_ir_gen_pass.impl/iterators.impl.jac:1-26`, `jac/tests/compiler/passes/native/test_native_iterators.jac`).

**Sketch.** Generators are the canonical state-machine lowering (as in C#, Rust, C++ coroutines), and the iterator ABI means they slot in without any new consumer-side protocol:

1. For a generator function `g`, emit an RC-managed frame struct `GenState_g = { resume_ip: i64, params..., cross-yield locals... }`. Only locals **live across a yield point** move into the frame; everything else stays as ordinary allocas in the resume function.
2. Calling `g(...)` allocates and initializes the frame (`resume_ip = 0`) and returns `JacIter { next_fn = __jac_gen_g_next, state = frame }`. Existing `for`-loops, `next()`, and the lazy adapters (`map` / `filter` / `enumerate` / `zip` take an upstream `JacIter`) compose with it unchanged.
3. `__jac_gen_g_next(frame*, out T*) -> i1` opens with a switch on `resume_ip` jumping to the block after each yield. `yield v` stores `v` to the out slot, records the next resume point, returns true. Function exit sets `resume_ip = DONE`, runs the frame drop, returns false.
4. RC integration: the frame gets a drop function releasing still-owned cross-yield locals (for early `break` / abandonment mid-iteration), registered the same way container release functions are (`_emit_container_release_fn` pattern), and the frame type registers with the cycle collector (a generator can capture a container that holds the generator).

Because the existing move-elision and region-arena machinery already exists (`test_move_elision.jac`, `test_region_arena.jac`), a non-escaping `for x in g(...)` frame is a natural later candidate for stack / arena placement, at which point simple generators optimize down to the equivalent loop. That is follow-on work, not a landing requirement.

**Scope.** Plain `yield expr`, including inside loops and conditionals. Excluded initially (each is separable): `yield from`, yield inside `try`/`finally`, the `send` / `throw` protocol, async generators. The checker side should type `g()` as the same `Iterator[T]` the #6403 work introduced.

**Acceptance.** Cross-backend equivalence fixtures: sum over a generator; early `break` then resume a second independent instance; generator passed to `map` / `filter`; nested generators; StopIteration contract via first-class `next()`; a cycle-collector test with a self-referential capture.

---

## Workstream E: closure captures for nested functions (RC cells)

**Current state.** Nested `def`s lower fine when they capture nothing; `get_enclosing_captures()` computes the free-variable set and any capture raises E5092 (`jac/.../na_ir_gen_pass.impl/func.impl.jac:448-459`, plus the guardrail test). Lambdas, by contrast, already capture, but **by value**: captured locals are snapshotted at creation and appended as bound parameters (`_codegen_lambda`, `expr.impl.jac:124`). That snapshot diverges from the Python backend's semantics whenever the captured variable is mutated after the lambda is created (Python closures observe the variable, not its value at capture time), so today's lambda behavior is subtly wrong under mutation, and this workstream fixes that too.

**Sketch: per-variable RC cells, with an analysis-justified by-value fast path.**

1. Classify each captured symbol of a closure (lambda or nested def):
   - **(a) Immutable after capture:** never assigned after any closure-creation point and never assigned inside a nested scope. By-value snapshot is observationally equivalent; keep the existing fast path, now justified by analysis instead of applied unconditionally.
   - **(b) Otherwise:** promote that variable's storage from an entry alloca to a heap cell `Cell_T = { rc header, value: T }`, allocated at the declaration site. All reads and writes in the enclosing function go through `cell->value`; the closure record holds an RC reference to the cell; writes from inside the closure store through the cell, which also yields correct nonlocal-write semantics.
2. Per-variable cells, not a shared frame box, are the deliberate long-term choice: a long-lived closure must not pin unrelated locals of its enclosing function.
3. The cost is one allocation per promoted variable plus one indirection on its accesses, confined to functions containing capturing closures. Programs without closures see zero change.
4. RC / cycles: cells are ordinary RC objects; closure-to-cell-to-closure cycles are collectable by registering cell types with the cycle collector via the existing container registration path.
5. Unify lambdas and nested defs on this one lowering, removing the E5092 and the current semantic divergence in the same change.

The capture sets already exist (`nd.captures` for lambdas, `get_enclosing_captures()` for nested defs); the only new analysis is the assigned-after-capture check, a single pass over the enclosing body comparing assignment positions against closure-creation positions.

**Acceptance.** Equivalence fixtures: `make_counter()` returning a mutating closure; the loop-capture late-binding case (all closures observe the final loop value, matching the Python backend); a nested def writing an enclosing local; a closure escaping by return and by storage into a list (cycle case); plus an IR-shape test asserting the by-value fast path still emits no cell for case (a).

---

## Cross-cutting

- **Ordering.** C0 (decorator guardrail) first, it is a small correctness fix. A and B are small and independent. E before D is the natural order: D's frame-drop machinery reuses the RC-struct-with-drop-fn shape E introduces, though neither blocks the other.
- **Capability gate hygiene.** As each feature lands, remove it from the reject path and add the note to `test_native_feature_guardrail.jac` (the same pattern used when `visit` / `disengage` / `report` graduated). `YieldExpr` and `LambdaExpr` should remain **auto-promotion disqualifiers** (`NATIVE_DISQUALIFIER_NODES`) initially even after their explicit-`.na.jac` support lands, and graduate separately once burned in.
- **The invariant to hold in review:** every one of these lowers to code that exists only at use sites. If a change to any workstream requires touching object layout, field-access codegen, or adding a check to a path shared by non-users of the feature, that change is wrong for this issue.

### PR DESCRIPTION
**Pull request by marsninja**
Closes #7588.

Implements all five workstreams of the zero-global-cost native gaps issue. Every feature lands test-first (each commit's suite was written and observed red before the lowering), all cost is confined to use sites, and no object representation or shared codegen path changed for programs that do not use these features.

## C0 + C1: decorators

- **Guardrail first (C0):** a decorated function, method, or archetype in a native module used to compile with the decorator silently dropped, running the undecorated body. The capability gate now rejects decoration in native contexts with E5090 anchored on the decorator expression, and decorated abilities/archetypes disqualify markerless modules from native auto-promotion.
- **Static decoration (C1):** module-level, non-generator functions whose decorators are plain names now lower: the body compiles under a private raw symbol, a private function-pointer slot is initialized at module glob-init by applying the decorator chain bottom-up (JacVal payload round-trip for `any`-typed decorators), and calls go through the slot. A decorator returning a different function genuinely redirects calls; registration decorators observe their glob-init side effects; stacking composes. Method/archetype/generator/dynamic decorators stay guardrailed.
- Tests: `tests/compiler/passes/native/test_native_decorators.jac` (+ `fixtures/decorators.na.jac`).

## A: os.path `join` / `abspath` in the bundled path kernel

- `join` (posixpath fold semantics via optional-sentinel trailing components) and `abspath` (`normpath(join(getcwd(), p))`, a pure string computation, unlike the existing symlink-resolving `realpath`) land in `na_stdlib/os/path.na.jac`, completing the bundled kernel alongside `split`/`splitext`/`normpath`.
- Landing this exposed a call-lowering hole affecting every bundled native module: a frontend signature wider than the native implementation (variadic stub over a bounded native function) built fewer IR args than the callee's arity and crashed CallInstr with IndexError. `_codegen_bound_args` now pads trailing slots with the callee's default fallback and reports a structured E5092 when a call genuinely exceeds native arity.
- Tests: `prim_ospath.jac` equivalence fixture extended (join edge cases incl. `join("a", "")`, absolute-component reset; abspath over absolute/relative/dot/empty).

## B: `isinstance` over mixed builtin + archetype tuples

- `isinstance(x, (int, Foo))` raised E5092 (the hole left between #6421 and #6932). It now lowers as a guarded disjunction on one boxed operand: JacVal tag compares for builtin arms, then only behind an object-tag, non-null-payload branch a vtable load compared against every type whose MRO contains an archetype arm, so subtypes match and non-pointer payloads are never dereferenced.
- The layout pass flags archetypes named as isinstance targets vtable-required (hierarchy members already are), covering standalone types; cost is confined to named targets and mixed call sites.
- The old E5092 guardrail graduates to `test_native_isinstance_mixed.jac`.

## E: nested-function closure captures

- Capturing nested defs were rejected wholesale. They now lower with the frame machinery the lambda path proved out, extended for Jac's implicit-rebind semantics: read-only captures forward as hidden value parameters at each call (late-bound, observationally identical to sv), and rebound captures (the frontend's `implicit_rebinds` set, the same source sv derives `nonlocal` from) forward as a pointer to the enclosing frame's slot so writes propagate and RC ownership stays with the one slot. Self-recursion forwards uniformly.
- Escape shapes (a capturing closure returned, stored, or passed as a value) would dangle frame slots, so codegen rejects them with a structured E5092, which also closes a pre-existing silent mis-binding hazard for escaping capturing lambdas. Two malformed-constant fallbacks (aggregate-typed return/default arg) now emit zeroinitializer instead of crashing the IR serializer.
- Tests: `test_native_closures.jac` (+ `fixtures/closures.na.jac`), cross-backend parity in `prim_closures.jac`.

## D: generators (`yield`) as JacIter state machines

- A `def` with `yield` (declared `-> Iterator[T]`-family) compiles to a creation function returning `JacIter{next_fn, frame}` over an RC-managed, zeroed frame (drop function hooked into the RC dtor slot; pointer params retained, the frame owns them) plus a private resume function `i1(i8* frame, i8* out)` dispatching on the frame's resume index. `yield` stores the element, stamps the next index, returns true; bare `return`/fall-off cleans up, stamps done, returns false; exhausted frames keep returning false. Consumers (`for`, first-class `next()` with the StopIteration contract, the lazy adapters) work unchanged because the ABI is the existing first-class iterator ABI from #6403.
- Resumability drives the design: inside a generator, `_entry_alloca` carves 8-byte-aligned slots out of the frame instead of stack allocas, so every local, loop index, and iterator temp survives suspension and resuming into the middle of a loop is sound. Loop lowerings that kept SSA across their bodies spill to frame slots when the body yields (protocol-loop iterator value; range stop/step); indexed iteration over materialized collections with a yielding body is a structured E5092 naming the `iter()`/while workaround. Abandoned generators release live pointer slots through the drop function.
- `map`/`filter` now also accept inline non-capturing lambdas as callbacks (previously named functions only).
- Guardrails: yield outside a generator body, `yield from`, bare yield, yield inside `try`, and `return <value>` in a generator are structured E5092s. `YieldExpr` stays a native auto-promotion disqualifier.
- Tests: `test_native_generators.jac` (+ `fixtures/generators.na.jac`), cross-backend parity in `prim_generators.jac`.

## Notes

- Pre-existing, unrelated: the strict py-side checker leaks `_T` for `next()` on a user generator call result and for nested `for` loops over generic-returning calls; the parity fixtures avoid those shapes and the native behavioral suites cover them instead.
- Targeted suites run per workstream: `test_native_decorators.jac`, `test_native_isinstance_mixed.jac`, `test_native_closures.jac`, `test_native_generators.jac`, `test_native_feature_guardrail.jac`, `test_native_iterators.jac`, `test_native_obj_str.jac`, `test_native_lexer.jac`, plus targeted `test_native_gen_pass.jac` slices (isinstance/class/nested/lambda/for/range/glob/osp/async) and the `test_prim_equivalence.jac` fixtures (`os.path bundled`, `closures nested capture`, `generators yield`). The 9 `test_osp_equivalence.jac` failures reproduce identically on a clean tree in this environment (known local-env issue) and are untouched by this branch.

### CODE REVIEW

File: jac/jaclang/compiler/passes/native/na_ir_gen_pass.impl/calls.impl.jac [hunk 1]
```diff
@@ -110,6 +113,51 @@ impl NaIRGenPass._codegen_bound_args(
...
+            self.emit(
+                E5092,
+                node_override=nd,
+                kind="expression",
+                name=f"call with {len(args)} arguments (native accepts {expected_argc})",
+                help_text="The resolved (frontend) signature allows more arguments than the native implementation declares (a variadic stub over a bounded native function). Pass fewer arguments or fold the extras before the call."
+            );
+            return args[:expected_argc];
+        }
+        return args;
```
[the remark this answers is not in the document]
Comment (marsninja) [reply jac/jaclang/compiler/passes/native/na_ir_gen_pass.impl/calls.impl.jac:156]: Deliberate, and empirically load-bearing: `func` here can be a name-keyed extern whose declared arity understates the real linked symbol. Module-qualified bundled calls hit this — `hmac.new(key, msg, digestmod)` resolves an arity-2 extern over the three-parameter native kernel, so the third argument must reach the symbol via the call ABI. The truncating version suggested here is exactly what commit fd020a143 tried, and it segfaulted the equivalence suite at runtime (digestmod never arrived); the pass-through matches main's long-standing behavior and the current run is green with it. The binding's CallInstr validates the declared parameters and emits the extras positionally, which the JIT accepts. Added a comment in the code documenting the constraint.
